### ISSUE TITLE
HEADs converted to GETs on relayed requests

### ISSUE THREAD

**Issue by jonathan-annett** [olalonde/follow-redirects#39]
HEADs where being converted to GETs, meaning HEADing a relayed server results in the entire payload being dumped unceremoniously on the requester.

I also had an issue with being stuck in an endless relay loop.

my quick and dirty hack to make it work can be cut and pasted to the top of any jsnode file, (don't worry with npm or other bollocks, just have assert,debug and consume installed.

enjoy. 

```
'use strict';

var assert = require('assert');
var debug = require('debug')('follow-redirects');
var consume = require('stream-consume');
// follow-redirects is taken and butchered from code obtained using
// "npm install follow-redirects"
// ie: https://www.npmjs.com/package/follow-redirects
// it had enough bugs in it to warrant fixing, but
// it's not worth doing
var followRedirects = (function (_nativeProtocols) {
  
  var nativeProtocols = {};

  var publicApi = {
    maxRedirects: 5
  };

  for (var p in _nativeProtocols) {
    /* istanbul ignore else */
    if (_nativeProtocols.hasOwnProperty(p)) {
      // http://www.ietf.org/rfc/rfc2396.txt - Section 3.1
      assert(/^[A-Z][A-Z\+\-\.]*$/i.test(p), JSON.stringify(p) + ' is not a valid scheme name');
      generateWrapper(p, _nativeProtocols[p]);
    }
  }

  return publicApi;

  // to redirect the result must have
  // a statusCode between 300-399
  // and a `Location` header
  function isRedirect(res) {
    return (res.statusCode >= 300 && res.statusCode <= 399 &&
    'location' in res.headers);
  }


  function execute(options, callback) {
    var req_method=options.method;
    var fetchedUrls = [];
    var clientRequest = cb();

    // return a proxy to the request with separate event handling
    var requestProxy = Object.create(clientRequest);
    requestProxy._events = {};
    requestProxy._eventsCount = 0;
    if (callback) {
      requestProxy.on('response', callback);
    }
    return requestProxy;

    function cb(res) {
      // skip the redirection logic on the first call.
      if (res) {
        var fetchedUrl = url.format(options);
        fetchedUrls.unshift(fetchedUrl);

        if (!isRedirect(res)) {
          res.fetchedUrls = fetchedUrls;
          requestProxy.emit('response', res);
          return;
        }

        // we are going to follow the redirect, but in node 0.10 we must first attach a data listener
        // to consume the stream and send the 'end' event
        consume(res);

        // need to use url.resolve() in case location is a relative URL
        var redirectUrl = url.resolve(fetchedUrl, res.headers.location);
        debug('redirecting to', redirectUrl);

        options=url.parse(redirectUrl);
      }

      if (fetchedUrls.length > options.maxRedirects) {
        var err = new Error('Max redirects exceeded.');
        return forwardError(err);
      }

      options.nativeProtocol = nativeProtocols[options.protocol];
      options.defaultRequest = defaultMakeRequest;

      var req = (options.makeRequest || defaultMakeRequest)(options, cb, res);
      req.on('error', forwardError);
      return req;
    }

    function defaultMakeRequest(options, cb, res) {
      if (res && res.statusCode !== 307) {
        // This is a redirect, so use only GET methods, except for status 307,
        // which must honor the previous request method.
        options.method = 'GET';
      }

      if (req_method === 'HEAD') 
        options.method=req_method;

      var req = options.nativeProtocol.request(options, cb);

      if (res) {
        // We leave the user to call `end` on the first request
        req.end();
      }

      return req;
    }

    // bubble errors that occur on the redirect back up to the initiating client request
    // object, otherwise they wind up killing the process.
    function forwardError(err) {
      requestProxy.emit('error', err);
    }
  }

  function generateWrapper(scheme, nativeProtocol) {
    var wrappedProtocol = scheme + ':';
    var H = function () {};
    H.prototype = nativeProtocols[wrappedProtocol] = nativeProtocol;
    H = new H();
    publicApi[scheme] = H;

    H.request = function (options, callback) {
      return execute(parseOptions(options, wrappedProtocol), callback);
    };

    // see https://github.com/joyent/node/blob/master/lib/http.js#L1623
    H.get = function (options, callback) {
      var req = execute(parseOptions(options, wrappedProtocol), callback);
      req.end();
      return req;
    };
  }

  // returns a safe copy of options (or a parsed url object if options was a string).
  // validates that the supplied callback is a function
  function parseOptions(options, wrappedProtocol) {
    if (typeof options === 'string') {
      options = url.parse(options);
      options.maxRedirects = publicApi.maxRedirects;
    } else {
      options.maxRedirects = publicApi.maxRedirects;
      options.protocol     = wrappedProtocol;
    }
    assert.equal(options.protocol, wrappedProtocol, 'protocol mismatch');

    debug('options', options);
    return options;
  }
}) ({
  http: require('http'),
  https: require('https')
});
followRedirects.maxRedirects = 10; 
var  http  = followRedirects.http,
     https = followRedirects.https;

```

**Comment by jamestalmage** [comment]
Yeah, it doesn't make any sense to forward a HEAD to a GET. PR welcome.

**Comment by RubenVerborgh** [comment]
Should be fixed along with #12.

**Comment by RubenVerborgh** [comment]
Fixed by cd203e6f4710059683fb9794db5a720a4fa4f1d1.
